docs/input_output.md

```diff
@@ -2,85 +2,279 @@
+The Scheduler service creates the initial event by combining the collection configuration, external resource configuration, workflow configuration, and deployment environment settings.  The subsequent workflow messages between tasks must conform to the message schema. By using Cumulus Message Adapter, individule task Lambda function can focus on the input and output specific to the task, and does not worry about the non-task related message fields.
 
-Because of the potential size of the `"payload"` field, it may contain a reference to an S3 Bucket and Key, as follows:
+Cumulus Message Adapter implements an AWS Lambda handler that adapts incoming messages in the Cumulus protocol to a format more easily consumable by Cumulus tasks, invokes the tasks, and then adapts their response back to the Cumulus message protocol to be sent to the next task.  
+
+A task Lambda function can be configured to use Cumulus Message Adapter to help construct the input/output messages and resolve task configurations. In the Lambda function configuration file, a task Lambda function can be configured to use Cumulus Message Adapter, for example,
+
+    DiscoverPdrs:
+      handler: index.handler
+      useMessageAdapter: true  
+
+A task Lambda function using the Message Adapter shall take a event message which contains two fields, `"input"` and `"config"`.
+> `input`: By default, the incoming payload is the payload from previous task, or it can be a portion of the payload as specified in the task configuration.
+
+> `config`: Task-specific configuration object with URL templates resolved.
+
+> `Task output`: By default, the task's return value is the next payload, or a portion of it is interpreted as the next payload as specified in the task configuration.
+
+**Cumulus Message Adapter has the following capabilities:**
+
+### 1. Retrieve a Cumulus message from S3 Bucket, or store a cumulus message to a S3 Bucket.
+
+Because of the potential size of a Cumulus message, mainly the `"payload"` field, if the message size is greater than 10000 bytes, the full message will be stored to S3 Bucket.  The message may contain a reference to an S3 Bucket and Key, as follows:
 
     {
-      "payload" : {
-        "Bucket" : "gitc-foo",
-        "Key" : "bar/baz"
+      "cumulus_meta": {...},
+      "replace": {
+        "Bucket": "gitc-foo",
+        "Key": "bar/baz"
+        }
+    }
+
+When a workflow task receives such message, the Cumulus Message Adapter is responsible for fetching the JSON message document and pass it to the task.  When a task output message is too big, the Cumulus Message Adapter will store the message to S3 Bucket under $.cumulus_meta.buckets.internal, and return a new message with S3 reference as above example.
```

Comment:
"When a workflow receives a large message,"
"is responsible for fetching the JSON message document and passing it to the task" 
"with an S3 reference as in the above example."